p8e-encryption/src/main/kotlin/io/provenance/p8e/encryption/ecies/ProvenanceECIESCipher.kt

```diff
@@ -0,0 +1,109 @@
+     * @throws IllegalArgumentException If ProvenanceECIESEncrypt has already been used..
+     * @throws IllegalArgumentException If Unable to generate an ephemeral key pair
+     */
+    @Throws(ProvenanceECIESEncryptException::class, IllegalArgumentException::class)
+    fun encrypt(data: ByteArray, publicKey: PublicKey, additionalAuthenticatedData: String?): ProvenanceECIESCryptogram {
+        try {
+            val ephemeralKeyPair = ProvenanceKeyGenerator.generateKeyPair(publicKey)
+
+            // Derive a secret key from the public key of the recipient and the ephemeral private key.
+            val ephemeralDerivedSecretKey = ProvenanceKeyGenerator.computeSharedKey(ephemeralKeyPair.private, publicKey).let {
+                ProvenanceHKDFSHA256.derive(it.encoded,null, ECUtils.KDF_SIZE)
+            }
+
+            // Encrypt the data
+            val encKeyBytes = Arrays.copyOf(ephemeralDerivedSecretKey, 32)
+            val encKey = ProvenanceAESCrypt.secretKeySpecGenerate(encKeyBytes)
+            val body = ProvenanceAESCrypt.encrypt(data, additionalAuthenticatedData, encKey)
+
+            // Compute MAC of the data
+            val macKeyBytes = Arrays.copyOfRange(ephemeralDerivedSecretKey, 32, 64)
+
+            val tag = ProvenanceAESCrypt.encrypt(plainText = macKeyBytes!!, additionalAuthenticatedData = additionalAuthenticatedData, key = encKey, useZeroIV = true)
+
+            // Return encrypted payload
+            return ProvenanceECIESCryptogram(ephemeralKeyPair.public!!, tag!!, body!!)
+        } catch (e: InvalidKeyException) {
+            logger.error("Invalid key exception", e)
+            throw ProvenanceECIESEncryptException("Decryption error occurred", e)
+        } catch (e: BadPaddingException) {
+            logger.error("Bad padding ", e)
+            throw ProvenanceECIESEncryptException("Decryption error occurred", e)
+        } catch (e: IllegalBlockSizeException) {
+            logger.error("Illegal block size ", e)
+            throw ProvenanceECIESEncryptException("Decryption error occurred", e)
+        }
+    }
+
+    /**
+     * Decrypt provided encrypted payload.
+     *
+     * @param payload {@code ProvenanceECIESCryptogram} Payload to be decrypted.
+     * @return Decrypted bytes.
+     * @throws ProvenanceECIESDecryptException In case decryption fails due to invalid key or invalid MAC value.
+     * @throws IllegalArgumentException In case decryption fails due to invalid life-cycle phase,
+     */
+    @Throws(ProvenanceECIESDecryptException::class)
+    fun decrypt(payload: ProvenanceECIESCryptogram, keyRef: KeyRef, additionalAuthenticatedData: String?): ByteArray {
+        try {
+
+            val ephemeralDerivedSecretKey = if(keyRef.type == SMARTKEY) {
+                // Create a transient security object out of the ephemeral public key from the payload.
+                val transientEphemeralSObj = payload.ephemeralPublicKey.toTransientSecurityObject()
```

Comment:
So this creates the transient security object, and it lasts just through the request in toAgreeKey on the next line of code?

Reply:
I'm assuming there is no way to make that in one request instead of two somehow?

Reply:
yea, TTL is during this session, don't think I can make one request out of two, need the transient SObj to exist before we can transform it into an AgreeKey

Reply:
I would assume if the auth token refreshed between these two calls, that would still be the same 'session'? 🤞

Reply:
yea, the refresh extends the session